todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java

```diff
@@ -0,0 +1,214 @@
+        checkNotNull(fragmentActivity);
+        TasksMvpController tasksMvpController = new TasksMvpController(fragmentActivity);
+        tasksMvpController.initTasksView();
+        return tasksMvpController;
+    }
+
+    /**
+     * Creates a controller for a task detail view for phones or tablets.
+     * @param fragmentActivity the context activity
+     * @return a TasksMvpController
+     */
+    public static TasksMvpController createTaskDetailView(
+            @NonNull FragmentActivity fragmentActivity, String taskId) {
+        checkNotNull(fragmentActivity);
+        TasksMvpController tasksMvpController = new TasksMvpController(fragmentActivity);
+        tasksMvpController.initTaskDetailView(taskId);
+        return tasksMvpController;
+    }
+
+    // Force factory method, prevent direct instantiation:
+    private TasksMvpController(@NonNull FragmentActivity fragmentActivity) {
+        mFragmentActivity = fragmentActivity;
+    }
+
+    private void initTasksView() {
+        if (isTablet()) {
+            createTabletElements();
+        } else {
+            TasksFragment tasksFragment = findOrCreateTasksFragment();
+            if (tasksFragment.getPresenter() != null) {
+                mTasksPresenter = (TasksPresenter) tasksFragment.getPresenter();
+            } else {
+                mTasksPresenter = createListPresenter(tasksFragment);
+            }
+            tasksFragment.setPresenter(mTasksPresenter);
+        }
+    }
+
+    private void initTaskDetailView(String taskId) {
+        TaskDetailFragment taskDetailFragment = findOrCreateTaskDetailFragment(taskId);
+        mTaskDetailPresenter = createTaskDetailPresenter(taskId, taskDetailFragment);
+        taskDetailFragment.setPresenter(mTaskDetailPresenter);
+    }
+
+    public TasksPresenter getTasksPresenter() {
+        return mTasksPresenter;
+    }
+
+    public TaskDetailPresenter getTaskDetailPresenter() {
+        return mTaskDetailPresenter;
+    }
+
+    /**
+     * To be called in tablet mode when a new detail view is requested.
+     * @param taskId the ID of the task
+     * @return the fragment to be added to the UI
+     */
+    public TaskDetailFragment createDetailViewForTablet(String taskId) {
+        // Create the View
+        TaskDetailFragment taskDetailFragment = TaskDetailFragment.newInstance(taskId);
+
+        // Create the Presenter
+        TaskDetailPresenter taskDetailPresenter = createTaskDetailPresenter(taskId,
+                taskDetailFragment);
+
+        // Wire presenters
+        mTasksTabletPresenter.setTaskDetailPresenter(taskDetailPresenter);
+        taskDetailFragment.setPresenter(mTasksTabletPresenter);
+        return taskDetailFragment;
+    }
+
+    private void createTabletElements() {
+        TasksFragment tasksFragment = findOrCreateTasksFragment();
+
+        // TasksFragment is retained so let's reuse its presenter.
+        if (tasksFragment.getPresenter() != null) {
```

Comment:
same as above, the `Fragment` should not expose the `Presenter`, I'd rather have this `Controller` in charge of that.